src/illumidesk/spawners/spawners.py

```diff
@@ -0,0 +1,166 @@
+import os
+import shutil
+
+from dockerspawner import DockerSpawner
+
+from illumidesk.authenticators.constants import WORKSPACE_TYPES
+
+
+class IllumiDeskBaseDockerSpawner(DockerSpawner):
+    """
+    Custom base DockerSpawner which provides the option to designate the image to
+    spawn based on values set within the authentication dictionary's ``auth_state``, ``nested-key``,
+    where the ``nested-key`` has a key name of your choosing.
+
+    For example, if you would like to spawn an image based on the nested key's value
+    ``foo`` the authenticator should return a dictionary with ``authentication['auth_state']['foo']``.
+    This base class requires that ``Authenticator.enable_auth_state = True.``
+    """
+
+    def _image_from_key(self, *kwargs: str) -> str:
```

Comment:
This is the generic method to set values based on key from auth_state. I renamed it from `_image_from_type` to `_image_from_key` since now its somewhat generic (images and workspaces use it now).